text/0000-cargo-report-future-incompat.md

```diff
@@ -0,0 +1,719 @@
+
+This RFC suggests the provided feedback take the form of a summary at
+the end of cargo's build of `unwary`, as illustrated in the explanation below.
+
+Furthermore, we want the feedback to provide guidance as to how the
+`unwary` maintainer can address the issue. Here are some potential
+forms this additional guidance could take.
+
+ * If cargo is not running in "offline mode", it can take the
+   future-incompatibilty signaling as an opportunity to query
+   `crates.io` to find out if a newer version of the upstream crate is
+   available, and if so, suggest to the user they might upgrade to it.
+   If such an upgrade could be done via `cargo update`, then the
+   output could obviously suggest that as well. (This is just a
+   heuristic measure, as it would not attempt to check ahead of time
+   if the newer version actually resolves the problem in question.)
+
+ * Cargo could suggest to the `unwary` maintainer that they file a bug
+   (or search for previously-filed bug) in the source repository for
+   the upstream crate that is issuing the future-incompatibility
+   warning. (That is, the `brash` author might not be aware of the
+   issue; for example, if they last updated their crate before the
+   lint in question was deployed on the Rust compiler.)
+
+ * `rustc` itself could embed, for each future-incompatibility lint,
+   how soon the Rust developers will turn the lint to a hard error.
+   This would give the `unwary` maintainer an idea of how much time
+   they have before they will be forced to address the issue (by
+   posting a PR upstream, or switching to a fork of `brash`, et
+   cetera).
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+After cargo finishes compiling a crate and its upstream dependencies,
+it may include a final warning about *future incompatibilties*.
+
+A future incompatibility is a pattern of code that is scheduled to be
+removed from the Rust language in some future release. Such code patterns
+are usually instances of constructs that can exhibit undefined behavior
+(i.e. they are unsound) or do not have a well-defined semantics,
+but are nonetheless in widespread use and thus need a
+grace period before they are removed.
+
+If any crate or any of its upstream dependencies has code that
+triggers a future incompatibility warning, but the overall compilation
+is otherwise without error, then cargo will report all instances of
+crates with future incompatibilities at the end of the compilation.
+When possible, this report includes the future date or release version
+where we expect Rust to stop compiling the code in question.
+
+Example:
+
+```
+crates % cd unwary
+unwary % cargo build
+   Compiling brash v0.1.0
+   Compiling bold v0.1.0
+   Compiling rash v0.1.0
+   Compiling unwary v0.1.0
+    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
+
+    warning: the crates brash, bold, and rash contain code that will be rejected by a future version of Rust.
```

Comment:
Nit: it should mention the version number (`brash v0.1.0`) because sometimes multiple versions of crates are contained in a crate DAG and only some versions could be affected.

Reply:
Will fix

Reply:
resolved (I think) by 90415f83f07985f4e883ef2bf75a03666f943b8c